Fix Axis.getDirection() to never return null (#80)

Manas-Dikshit opened the pull request:
This pull request addresses issue #80, where Axis.getDirection() could return null if the underlying Property.DIRECTION was missing or unrecognized. Returning null violates the CoordinateSystemAxis GeoAPI contract and can lead to NullPointerException in downstream code.

Approach:
Updated the getDirection() method in Axis.java to check the result of the internal search method.
If search returns null, the method now returns the default value AxisDirection.OTHER.
Preserved existing behavior for recognized axis directions.

How It Solves the Problem:
Ensures getDirection() always returns a non-null AxisDirection, maintaining compliance with GeoAPI expectations.
Prevents runtime crashes and makes PROJ-JNI safer for all consumers of CRS objects.
No breaking changes for existing valid axis directions.

Example:
AxisDirection direction = axis.getDirection(); // Never null, defaults to OTHER if unknown

Impact:
Fixes null-safety issue in Axis class.
Improves robustness and GeoAPI compliance.

Manas-Dikshit committed:
Update Axis.java

desruisseaux commented:
Do you have the Well Known Text (WKT) definition of the Coordinate Reference System (CRS) for which the axis direction is null? It would be for checking if we can infer a better axis direction than `OTHER` at least for the case that you encountered.

Manas-Dikshit commented:
Thank you for the feedback, @desruisseaux!

Here’s the WKT of the CRS where AxisDirection was null:

' ' ' wkt
GEOGCRS["Unknown datum based CRS",
    DATUM["Unknown based datum", ELLIPSOID["GRS 1980",6378137,298.257222101]],
    CS[ellipsoidal,2],
    AXIS["Geodetic latitude (Lat)", north, ANGLEUNIT["degree",0.0174532925199433]],
    AXIS["Geodetic longitude (Lon)", east, ANGLEUNIT["degree",0.0174532925199433]]
]
' ' '
It seems that when the Property.DIRECTION property was missing for one of the axes, the JNI wrapper didn’t resolve the direction string. Returning AxisDirection.OTHER prevents a NullPointerException, but if a consistent pattern is detectable from WKT (e.g., latitude → NORTH, longitude → EAST), we could consider inferring those directions as an enhancement.

desruisseaux commented:
While latitude is very likely to be oriented toward North, we cannot be sure as it could also be South. Same for longitude. It is safer to not try to guess.

I do not know if the issue is on PROJ side or elsewhere. On one side, PROJ-JNI is implemented only as below, relying fully on PROJ:

```cpp
value = get_shared_object<CoordinateSystemAxis>(env, object)->direction().toString().c_str();
```

On the other side, PROJ seems to have the information somewhere since it appears in the WKT. I suggest to keep the pull request as-is with only one amendment (I will post a comment directly on the line of code). Maybe the issue will disappear in some future PROJ version.

desruisseaux commented on src/main/java/org/osgeo/proj/Axis.java:
```diff
@@ -69,7 +69,9 @@ public String getAbbreviation() {
-        return search(AxisDirection.class, dir);
+        AxisDirection direction = search(AxisDirection.class, dir);
+        // Ensure a non-null return; default to OTHER if unknown
+        return (direction != null) ? direction : AxisDirection.OTHER;
```
Replace `AxisDirection.OTHER` by `AxisDirection.valueOf("UNSPECIFIED")` and add the following comment:

```
// TODO: Use AxisDirection.UNSPECIFIED in GeoAPI 3.1.
```

The "Unspecified" axis direction has been added in the ISO 19111:2019 international standard. GeoAPI has not yet been updated for that revision, but this is planed for GeoAPI 3.1. In the meantime, the use of `valueOf(String)` should allow us to anticipate.

Manas-Dikshit committed:
Update getDirection to handle null values

Ensure a non-null return for axis direction, defaulting to UNSPECIFIED if unknown.

Manas-Dikshit committed:
Merge branch 'main' into patch-2

Manas-Dikshit committed:
Merge pull request #1 from Manas-Dikshit/patch-2

Update getDirection to handle null values

Manas-Dikshit commented:
Thanks, @desruisseaux — updated accordingly.
The getDirection() method now uses AxisDirection.valueOf("UNSPECIFIED") with the TODO comment referencing GeoAPI 3.1 and ISO 19111:2019, as suggested.

This approach should maintain compatibility with current GeoAPI versions while anticipating the addition of UNSPECIFIED. The logic still guarantees a non-null return value, ensuring downstream safety.

desruisseaux commented on src/main/java/org/osgeo/proj/Axis.java:
```diff
@@ -68,18 +68,26 @@ public String getAbbreviation() {
     public AxisDirection getDirection() {
-        final String dir = impl.getStringProperty(Property.DIRECTION);
-        return search(AxisDirection.class, dir);
+    final String dir = impl.getStringProperty(Property.DIRECTION);
```
Fix indentation.

desruisseaux commented on src/main/java/org/osgeo/proj/Axis.java:
```diff
@@ -68,18 +68,26 @@ public String getAbbreviation() {
+    // GeoAPI has not yet been updated for that revision, but this change anticipates it.
     }
 
+
```
Omit those extra blank lines (only one blank line between methods).

Manas-Dikshit committed:
Update Axis.java

Manas-Dikshit commented:
@desruisseaux 
Omitted that extra blank line & also fixed indentation.

Manas-Dikshit committed:
Refactor getDirection method for clarity

desruisseaux commented:
Merged, thanks!
